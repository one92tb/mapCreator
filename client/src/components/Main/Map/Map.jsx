/*global google*/
import React, { Component } from "react";
import { connect } from "react-redux";
import { postSelectedMarker } from "../../../actions/postSelectedMarker";
import { fetchSelectedMarkers } from "../../../actions/fetchSelectedMarkers";
import { removeSelectedMarker } from "../../../actions/removeSelectedMarker";
import apiKey from "./apiKey";
import styled from "styled-components";
const { InfoBox } = require("react-google-maps/lib/components/addons/InfoBox");
const {
  compose,
  withProps,
  withStateHandlers,
  lifecycle
} = require("recompose");
const _ = require("lodash");
const {
  withScriptjs,
  withGoogleMap,
  GoogleMap,
  Marker,
} = require("react-google-maps");
const {
  SearchBox
} = require("react-google-maps/lib/components/places/SearchBox");

const Wrapper = styled.div`
  height: 100%;
  width: 100%;
  padding: 40px 20px 40px 10px;
  background: #f2f2f2;
`;

const SearchBoxInput = styled.input`
  box-sizing: border-box;
  border: 1px solid transparent;
  width: 240px;
  height: 40px;
  margin-top: 10px;
  padding: 0 12px;
  border-radius: 3px;
  box-shadow: 0 2px 6px rgba(0, 0, 0, 0.3);
  outline: none;
`;

const InfoBoxContainer = styled.div`
  display: flex;
  flex-direction: column;
  position: relative;
  background: #4ddbff;
  opacity: 0.9;
  height: 200px;
  width: 250px;
  border: 5px solid #00b8e6;
  border-radius: 5px;

  &:before {
    top: 100%;
    left: 50%;
    border: solid transparent;
    content: " ";
    height: 0;
    width: 0;
    position: absolute;
    pointer-events: none;
    border-color: rgba(255, 31, 150, 0);
    border-top-color: #00b8e6;
    border-width: 25px;
    margin-left: -25px;
  }
  &:after {
    top: 100%;
    left: 50%;
    border: solid transparent;
    content: " ";
    height: 0;
    width: 0;
    position: absolute;
    pointer-events: none;
    border-color: rgba(136, 183, 213, 0);
    border-top-color: #00b8e6;
    border-width: 20px;
    margin-left: -20px;
  }
`;

const InfoBtn = styled.button`
  margin-top: 10px;
  width: 160px;
  display: block;
  margin-left: auto;
  margin-right: auto;
  background: #4ddbff;
  color: #003d4d;
  border: 0;
  padding: 5px;
  border-radius: 2px;
  border: 2px solid #00b8e6;
  font-weight: bold;

  &:hover {
    cursor: pointer;
    color: #fff;
    background-color: #ff1a1a;
  }
`;

const InfoContent = styled.span`
  font-size: 15px;
  color: #003d4d;
  height: 35px;
  display: flex;
  align-items: center;
  justify-content: center;
  text-transform: uppercase;
  border-bottom: 2px solid #00b8e6;
  font-weight: bold;
`;

const onToggleOpen = ({ isOpen, id }) => id =>
  id && !isOpen
    ? {
        isOpen: !isOpen,
        id
      }
    : { id };

const geocode = (marker, postSelectedMarker) => {
  let geocoder = new window.google.maps.Geocoder();
  geocoder.geocode(
    {
      location: {
        lat: marker.lat,
        lng: marker.lng
      }
    },
    (results, status) => {
      console.log(results);
      if (status === "OK") {
        const markerToPost = {
          name: marker.name,
          lat: marker.lat,
          lng: marker.lng,
          icon: marker.icon,
          street: results[0].formatted_address.split(",")[0],
          city: results[0].formatted_address.split(",")[1],
          country: results[0].formatted_address.split(",")[2]
        };

        postSelectedMarker(markerToPost);
      } else if (status === "OVER_QUERY_LIMIT") {
        console.log(
          "Geocode was not successful for the following reason: " + status
        );
      } else {
        console.log(
          "Geocode was not successful for the following reason: " + status
        );
      }
    }
  );
};

const MapWithAMakredInfoWindow = compose(
  withProps({
    googleMapURL: `https://maps.googleapis.com/maps/api/js?key=${apiKey}&v=3.exp&libraries=geometry,drawing,places`,
    containerElement: (
      <div
        style={{
          height: `100%`
        }}
      />
    ),
    loadingElement: (
      <div
        style={{
          height: `100%`
        }}
      />
    ),
    mapElement: (
      <div
        style={{
          height: `100%`
        }}
      />
    )
  }),
  withStateHandlers(
    () => ({
      isOpen: false,
      id: null,
      bounds: null,
      center: {
        lat: 50.89973,
        lng: 15.72899
      },
      zoom: 12
    }),
    { onToggleOpen, geocode }
  ),
  withScriptjs,
  withGoogleMap,
  lifecycle({
    componentDidMount() {
      const refs = {};

      this.setState({
        onMapMounted: ref => {
          console.log(ref);
          refs.map = ref;
        },
        onBoundsChanged: () => {
          this.setState({
            bounds: refs.map.getBounds(),
            center: refs.map.getCenter()
          });
        },
        onSearchBoxMounted: ref => {
          refs.searchBox = ref;
        },
        onMouseOver: () => {},
        onPlacesChanged: () => {
          const places = refs.searchBox.getPlaces();
          const bounds = new google.maps.LatLngBounds();

          console.log(places, bounds);

          places.forEach(
            place =>
              place.geometry.viewport
                ? bounds.union(place.geometry.viewport)
                : bounds.extend(place.geometry.location)
          );

          const nextCenter = _.get(this.state.center);

          this.setState({
            center: nextCenter
          });

          refs.map.fitBounds(bounds);
        },
        onZoomChanged: () => {
          this.setState({
            zoom:
              refs.map.context.__SECRET_MAP_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.zoom
          });
        }
      });
    }
  })
)(props => {
  return (
    <GoogleMap
      ref={props.onMapMounted}
      onBoundsChanged={props.onBoundsChanged}
      onMouseOver={props.onMouseOver}
      defaultZoom={props.zoom}
      onZoomChanged={props.onZoomChanged}
      onClick={e => props.addMarker(e, geocode, postSelectedMarker)}
      defaultCenter={props.center}
      defaultCursor={props.cursor}
    >
      {props.markers
        .filter(
          marker =>
            !props.disableMarkers.find(
              disableItem => disableItem.name === marker.name
            )
        )
        .map((marker, index) => {
          return (
              <Marker
                onClick={() => props.onToggleOpen(marker.id)}
                key={index}
                position={{
                  lat: marker.lat,
                  lng: marker.lng
                }}
                icon={{
                  url: `http://localhost:8080/images/${marker.icon}`,
                  scaledSize: {
                    width: props.zoom < 14 ? 16 : 50,
                    height: props.zoom < 14 ? 16 : 50
                  }
                }}
                visible={!(props.zoom < 11)}
              >
                {props.isOpen &&
                  props.id === marker.id && (
                    <InfoBox
                      defaultPosition={{
                        lat: marker.lat,
                        lng: marker.lng
                      }}
                      ref={props.onInfoBox}
                      onCloseClick={props.onToggleOpen}
                      closeBoxURL=""
                      options={{
                        boxStyle: {
                          border: "none",
                          fontSize: "12pt",
                          overflow: "hidden",
                          height: "250px",
                          width: "250px",
                          display: props.zoom < 11 ? "none" : "block"
                        },
                        closeBoxMargin: "5px 5px 2px 2px",
                        alignBottom: true,
                        isHidden: false,
                        pixelOffset:
                          props.zoom < 14
                            ? new google.maps.Size(-125, -20)
                            : new google.maps.Size(-125, -45),
                        enableEventPropagation: true,
                        infoBoxClearance: new google.maps.Size(1, 1)
                      }}
                    >
                      <InfoBoxContainer>
                        <InfoContent>{marker.name}</InfoContent>
                        <InfoContent>{marker.street}</InfoContent>
                        <InfoContent>{marker.city}</InfoContent>
                        <InfoContent>{marker.country}</InfoContent>
                        <InfoBtn onClick={() => props.removeMarker(marker.id)}>
                          Remove Marker
                        </InfoBtn>
                      </InfoBoxContainer>
                    </InfoBox>
                  )}
              </Marker>
          );
        })}

      <SearchBox
        ref={props.onSearchBoxMounted}
        bounds={props.bounds}
        controlPosition={google.maps.ControlPosition.TOP_LEFT}
        onPlacesChanged={props.onPlacesChanged}
      >
        <SearchBoxInput type="text" placeholder="Customized your placeholder" />
      </SearchBox>
    </GoogleMap>
  );
});

class Map extends Component {
  componentDidMount() {
    this.props.fetchSelectedMarkers();
  }

  addMarker = (event, geocode) => {
    console.log(this.props);
    const selectedMarker = this.props.selectedMarker;
    const postSelectedMarker = this.props.postSelectedMarker;
    const isNavSelect = this.props.isNavSelect;

    if (selectedMarker.id && !selectedMarker.isDeleted && isNavSelect) {
      console.log(this.props.isNavSelect);
      const marker = {
        name: selectedMarker.name,
        icon: selectedMarker.icon,
        lat: event.latLng.lat(),
        lng: event.latLng.lng()
      };

      geocode(marker, postSelectedMarker);
    }
  };

  removeMarker = id => {
    this.props.removeSelectedMarker(id);
  };

  render() {
    console.log(this.props, this.state);
    return (
      <Wrapper>
        <MapWithAMakredInfoWindow
          markers={this.props.markers}
          selectedMarker={this.props.selectedMarker}
          addMarker={this.addMarker}
          removeMarker={this.removeMarker}
          disableMarkers={this.props.disableMarkers}
        />
      </Wrapper>
    );
  }
}

const mapStateToProps = state => ({
  selectedMarker: state.selectedMarker.selectedMarker,
  markers: state.selectedMarker.selectedMarkers,
  isNavSelect: state.selectedMarker.isNavSelect,
  disableMarkers: state.selectedMarker.disableMarkers
});

const mapDispatchToProps = {
  postSelectedMarker,
  fetchSelectedMarkers,
  removeSelectedMarker
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Map);
