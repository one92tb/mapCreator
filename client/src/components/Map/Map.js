import React, {Component} from 'react';
import GoogleMapReact from 'google-map-react';

const AnyReactComponent = () => <div><i className="icon-home"></i></div>;

class Map extends Component {

  constructor(props){
    super(props);
  }

  componentDidMount() {
    fetch('https://maps.googleapis.com/maps/api/js?key=AIzaSyC9KF6ohkrQU2JPS8wiBWLmlcUqbAf0ik8&callback=initMap')
    .then(data => console.log(data)) // JSON from `response.json()` call
    .catch(error => console.error(error));
  }

  static defaultProps = {
    center: {
      lat: 50.89973,
      lng: 15.72899
    },
    zoom: 12
  };

  render() {
    console.log(this.props);
    return (
      // Important! Always set the container height explicitly
      <div style={{ height: '100%', width: '100%' }}>
        <GoogleMapReact onClick={this.check}
          bootstrapURLKeys={{ key: 'AIzaSyC9KF6ohkrQU2JPS8wiBWLmlcUqbAf0ik8' }}
          defaultCenter={this.props.center}
          defaultZoom={this.props.zoom}
        >
          <AnyReactComponent
            lat={50.89973}
            lng={15.72899}
          />
        </GoogleMapReact>
      </div>
    );
  }

}

export default Map;
