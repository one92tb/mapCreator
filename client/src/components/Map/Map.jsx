import React, {Component} from 'react';
import {withScriptjs, withGoogleMap, GoogleMap, Marker, InfoWindow} from "react-google-maps"
import {connect} from 'react-redux';
import {postSelectedMarker} from '../../actions/postSelectedMarker';
import {fetchSelectedMarkers} from '../../actions/fetchSelectedMarkers';

const {compose, withProps, withStateHandlers} = require("recompose");

class Map extends Component {

  constructor(props) {
    super(props);
    this.state = {
      showingInfoWindow: false,
      activeMarker: {},
      selectedPlace: {},
      position: null
    };

    this.checkLocalization = this.checkLocalization.bind(this);
  }

  componentDidMount() {
    this.props.fetchSelectedMarkers();
  }

  static defaultProps = {
    center: {
      lat: 50.89973,
      lng: 15.72899
    },
    zoom: 12
  };

  checkLocalization(event) {
    console.log(event.latLng.lat(), event.latLng.lng());
    const selectedMarker = this.props.selectedMarker;
    const postSelectedMarker = this.props.postSelectedMarker;

    if (selectedMarker !== null) {
      const marker = {
        name: selectedMarker.name,
        icon: selectedMarker.icon,
        lat: event.latLng.lat(),
        lng: event.latLng.lng()
      }
      console.log(marker);
      postSelectedMarker(marker);
    }
  }

  render() {
    const markers = this.props.markers;
    const MapWithAMakredInfoWindow = compose(withStateHandlers(() => ({isOpen: false, id: null}), {
      onToggleOpen: ({isOpen, id}) => (id) => (id !== undefined && isOpen === false)
        ? ({
          isOpen: !isOpen,
          id: id
        })
        : ({id: id})
    }), withScriptjs, withGoogleMap)(props => {
      console.log(props);
      return <GoogleMap onClick={this.checkLocalization} defaultZoom={12} defaultCenter={{
          lat: 50.89973,
          lng: 15.72899
        }}>
        {
          markers.map((marker, index) => <Marker onClick={() => props.onToggleOpen(marker.id)} key={index} position={{
              lat: marker.lat,
              lng: marker.lng
            }} icon={{
              url: `http://localhost:8080/images/${marker.icon}`
            }}>
            {
              props.isOpen && props.id === marker.id && <InfoWindow onCloseClick={props.onToggleOpen}>
                  <h1>sss</h1>
                </InfoWindow>
            }
          </Marker>)
        }
      </GoogleMap>
    });

    return (<MapWithAMakredInfoWindow googleMapURL="https://maps.googleapis.com/maps/api/js?key=[API_KEY]&v=3.exp&libraries=geometry,drawing,places" 
      loadingElement={<div style = {{ height: `100%` }}/>}
      containerElement={<div style = {{ height: `100%` }}/>}
      mapElement={<div style = {{ height: `100%` }}/>}/>);
  }
}

const mapStateToProps = (state) => ({selectedMarker: state.selectedMarker.selectedMarker, markers: state.selectedMarker.selectedMarkers})

const mapDispatchToProps = {
  postSelectedMarker,
  fetchSelectedMarkers
}

export default connect(mapStateToProps, mapDispatchToProps)(Map);
//  onClick={() => this.onMarkerClick(marker.id)}
//onClick={props.onToggleOpen}
