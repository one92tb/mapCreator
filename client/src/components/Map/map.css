.card {
  width: 100%;
}
.markerStreet, .markerName {
  font-size: 15px;
  color: #fff;
  height: 35px;
  display: flex;
  align-items: center;
  justify-content: center;
  text-transform: uppercase;
    border-bottom: 2px solid #0052cc;
    font-weight: bold;
}

.infoBtn {
  margin-top: 10px;
  display: block;
  margin-left: auto;
  margin-right: auto;
  background: #0052cc;
  color: #fff;
  border: 0;
  padding: 5px;
  border-radius: 2px;
  border: 2px solid #003380;
  font-weight: bold;
}

.infoBtn:hover{
  cursor: pointer;
  background-color: #ff1a1a;
}
.infoBoxContainer {

  display: flex;
  flex-direction: column;
	position: relative;
	background: #1a75ff;
  opacity: 0.85;
  height: 200px;
  width: 250px;
  border: 5px solid #003380;
  border-radius: 5px;

}
.infoBoxContainer:after, .infoBoxContainer:before {
	top: 100%;
	left: 50%;
	border: solid transparent;
	content: " ";
	height: 0;
	width: 0;
	position: absolute;
	pointer-events: none;
}
.infoBoxContainer:after {
  border-color: rgba(136, 183, 213, 0);
	border-top-color: #1a75ff;
	border-width: 20px;
	margin-left: -20px;
}
.infoBoxContainer:before{
	border-color: rgba(255, 31, 150, 0);
	border-top-color: #003380;
	border-width: 25px;
	margin-left: -25px;
}
.searchBox{
  box-sizing: border-box;
  border: 1px solid transparent;
  width: 240px;
  height: 40px;
  margin-top: 10px;
  padding: 0 12px;
  border-radius: 3px;
  box-shadow: 0 2px 6px rgba(0, 0, 0, 0.3);
  outline: none;
}
