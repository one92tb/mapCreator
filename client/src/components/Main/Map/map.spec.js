import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import { shallowToJson } from "enzyme-to-json";
import configureStore from "redux-mock-store";
import { Map, MapWithAMakredInfoWindow } from "./Map";
import ShallowRenderer from "react-test-renderer/shallow";

const {
  withScriptjs,
  withGoogleMap,
  GoogleMap,
  Marker
} = require("react-google-maps");

const { InfoBox } = require("react-google-maps/lib/components/addons/InfoBox");
const {
  compose,
  withProps,
  withStateHandlers,
  lifecycle
} = require("recompose");

const _ = require("lodash");

describe("map component", () => {
  let wrapper;

  it("passess all props to map component", () => {
    const fetchIndicators = jest.fn();
    const postIndicator = jest.fn();
    const removeIndicator = jest.fn();
    const selectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1
    };
    const indicators = [
      {
        city: "Jelenia G贸ra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 2,
        lat: 50.92841093800815,
        lng: 15.640584390869208,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Jelenia G贸ra 58-500",
        country: " Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      }
    ];
    const disableMarkers = [];
    const selectedIndicator = "";
    const isNavSelect = true;

    wrapper = shallow(
      <Map
        fetchIndicators={fetchIndicators}
        postIndicator={postIndicator}
        removeIndicator={removeIndicator}
        indicators={indicators}
        isNavSelect={isNavSelect}
        selectedMarker={selectedMarker}
        disableMarkers={disableMarkers}
        selectedIndicator={selectedIndicator}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("passess all props to MapWithAMakredInfoWindow component", () => {
    const props = {
      containerElement: (
        <div
          style={{
            height: `100%`
          }}
        />
      ),
      loadingElement: (
        <div
          style={{
            height: `100%`
          }}
        />
      ),
      mapElement: (
        <div
          style={{
            height: `100%`
          }}
        />
      ),
      fetchIndicators: jest.fn(),
      postIndicator: jest.fn(),
      removeIndicator: jest.fn(),
      selectedMarker: {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      indicators: [
        {
          city: "Jelenia G贸ra 58-500",
          country: "Polska",
          icon: "1548784752159.png",
          id: 2,
          lat: 50.92841093800815,
          lng: 15.640584390869208,
          name: "test1",
          street: "Unnamed Road",
          userId: 1
        },
        {
          city: "Jelenia G贸ra 58-500",
          country: " Polska",
          icon: "1548784752159.png",
          id: 3,
          lat: 50.933387715648635,
          lng: 15.840054910888739,
          name: "test1",
          street: "Unnamed Road",
          userId: 1
        }
      ],
      disableMarkers: [],
      selectedIndicator: "",
      isNavSelect: true
    };

    const renderer = new ShallowRenderer();
    renderer.render(<MapWithAMakredInfoWindow {...props} />);
    const result = renderer.getRenderOutput();
    console.log(result);
    expect(result).toMatchSnapshot();
  });
});
