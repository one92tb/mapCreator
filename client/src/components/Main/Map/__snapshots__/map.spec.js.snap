// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map component passess all props to MapWithAMakredInfoWindow component 1`] = `
<withStateHandlers(withScriptjs(withGoogleMap(lifecycle(Component))))
  containerElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  disableMarkers={Array []}
  fetchIndicators={[Function]}
  googleMapURL="https://maps.googleapis.com/maps/api/js?key=AIzaSyB3kGdvwWFXzY4O0sg-pBzwFae51hFfguQ&v=3.exp&libraries=geometry,drawing,places"
  indicators={
    Array [
      Object {
        "city": "Jelenia G贸ra 58-500",
        "country": "Polska",
        "icon": "1548784752159.png",
        "id": 2,
        "lat": 50.92841093800815,
        "lng": 15.640584390869208,
        "name": "test1",
        "street": "Unnamed Road",
        "userId": 1,
      },
      Object {
        "city": "Jelenia G贸ra 58-500",
        "country": " Polska",
        "icon": "1548784752159.png",
        "id": 3,
        "lat": 50.933387715648635,
        "lng": 15.840054910888739,
        "name": "test1",
        "street": "Unnamed Road",
        "userId": 1,
      },
    ]
  }
  isNavSelect={true}
  loadingElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  mapElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  postIndicator={[Function]}
  removeIndicator={[Function]}
  selectedIndicator=""
  selectedMarker={
    Object {
      "icon": "1548784752159.png",
      "id": 1,
      "name": "test1",
      "userId": 1,
    }
  }
/>
`;

exports[`map component passess all props to map component 1`] = `
<div>
  <withProps(withStateHandlers(withScriptjs(withGoogleMap(lifecycle(Component)))))
    addIndicator={[Function]}
    disableMarkers={Array []}
    indicators={
      Array [
        Object {
          "city": "Jelenia G贸ra 58-500",
          "country": "Polska",
          "icon": "1548784752159.png",
          "id": 2,
          "lat": 50.92841093800815,
          "lng": 15.640584390869208,
          "name": "test1",
          "street": "Unnamed Road",
          "userId": 1,
        },
        Object {
          "city": "Jelenia G贸ra 58-500",
          "country": " Polska",
          "icon": "1548784752159.png",
          "id": 3,
          "lat": 50.933387715648635,
          "lng": 15.840054910888739,
          "name": "test1",
          "street": "Unnamed Road",
          "userId": 1,
        },
      ]
    }
    removeMarker={[Function]}
    selectedIndicator=""
    selectedMarker={
      Object {
        "icon": "1548784752159.png",
        "id": 1,
        "name": "test1",
        "userId": 1,
      }
    }
  />
</div>
`;
